// hono/dist/cjs/jsx/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var jsx_exports = {};
__export(jsx_exports, {
  ErrorBoundary: () => import_components.ErrorBoundary,
  Fragment: () => import_base.Fragment,
  Suspense: () => import_streaming.Suspense,
  cloneElement: () => import_base.cloneElement,
  createContext: () => import_context.createContext,
  createElement: () => import_base.jsx,
  default: () => jsx_default,
  isValidElement: () => import_base.isValidElement,
  jsx: () => import_base.jsx,
  memo: () => import_base.memo,
  startTransition: () => import_hooks.startTransition,
  startViewTransition: () => import_hooks.startViewTransition,
  use: () => import_hooks.use,
  useCallback: () => import_hooks.useCallback,
  useContext: () => import_context.useContext,
  useDebugValue: () => import_hooks.useDebugValue,
  useDeferredValue: () => import_hooks.useDeferredValue,
  useEffect: () => import_hooks.useEffect,
  useId: () => import_hooks.useId,
  useLayoutEffect: () => import_hooks.useLayoutEffect,
  useMemo: () => import_hooks.useMemo,
  useReducer: () => import_hooks.useReducer,
  useRef: () => import_hooks.useRef,
  useState: () => import_hooks.useState,
  useTransition: () => import_hooks.useTransition,
  useViewTransition: () => import_hooks.useViewTransition
});
module.exports = __toCommonJS(jsx_exports);
var import_base = require("./base");
var import_components = require("./components");
var import_context = require("./context");
var import_hooks = require("./hooks");
var import_streaming = require("./streaming");
__reExport(jsx_exports, require("./types"), module.exports);
var jsx_default = {
  memo: import_base.memo,
  Fragment: import_base.Fragment,
  isValidElement: import_base.isValidElement,
  createElement: import_base.jsx,
  cloneElement: import_base.cloneElement,
  ErrorBoundary: import_components.ErrorBoundary,
  createContext: import_context.createContext,
  useContext: import_context.useContext,
  useState: import_hooks.useState,
  useEffect: import_hooks.useEffect,
  useRef: import_hooks.useRef,
  useCallback: import_hooks.useCallback,
  useReducer: import_hooks.useReducer,
  useId: import_hooks.useId,
  useDebugValue: import_hooks.useDebugValue,
  use: import_hooks.use,
  startTransition: import_hooks.startTransition,
  useTransition: import_hooks.useTransition,
  useDeferredValue: import_hooks.useDeferredValue,
  startViewTransition: import_hooks.startViewTransition,
  useViewTransition: import_hooks.useViewTransition,
  useMemo: import_hooks.useMemo,
  useLayoutEffect: import_hooks.useLayoutEffect,
  Suspense: import_streaming.Suspense
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ErrorBoundary,
  Fragment,
  Suspense,
  cloneElement,
  createContext,
  createElement,
  isValidElement,
  jsx,
  memo,
  startTransition,
  startViewTransition,
  use,
  useCallback,
  useContext,
  useDebugValue,
  useDeferredValue,
  useEffect,
  useId,
  useLayoutEffect,
  useMemo,
  useReducer,
  useRef,
  useState,
  useTransition,
  useViewTransition
});
