// hono/dist/cjs/jsx/dom/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var dom_exports = {};
__export(dom_exports, {
  ErrorBoundary: () => import_components.ErrorBoundary,
  Suspense: () => import_components.Suspense,
  cloneElement: () => cloneElement,
  createContext: () => import_context2.createContext,
  createElement: () => createElement,
  default: () => dom_default,
  isValidElement: () => import_base.isValidElement,
  jsx: () => createElement,
  memo: () => import_base.memo,
  render: () => import_render.render,
  startTransition: () => import_hooks.startTransition,
  startViewTransition: () => import_hooks.startViewTransition,
  use: () => import_hooks.use,
  useCallback: () => import_hooks.useCallback,
  useContext: () => import_context.useContext,
  useDebugValue: () => import_hooks.useDebugValue,
  useDeferredValue: () => import_hooks.useDeferredValue,
  useEffect: () => import_hooks.useEffect,
  useId: () => import_hooks.useId,
  useLayoutEffect: () => import_hooks.useLayoutEffect,
  useMemo: () => import_hooks.useMemo,
  useReducer: () => import_hooks.useReducer,
  useRef: () => import_hooks.useRef,
  useState: () => import_hooks.useState,
  useTransition: () => import_hooks.useTransition,
  useViewTransition: () => import_hooks.useViewTransition
});
module.exports = __toCommonJS(dom_exports);
var import_base = require("../base");
var import_context = require("../context");
var import_hooks = require("../hooks");
var import_components = require("./components");
var import_context2 = require("./context");
var import_jsx_runtime = require("./jsx-runtime");
var import_render = require("./render");
const createElement = (tag, props, ...children) => {
  const jsxProps = { ...props, children };
  let key = void 0;
  if ("key" in jsxProps) {
    key = jsxProps.key;
    delete jsxProps.key;
  }
  return (0, import_jsx_runtime.jsx)(tag, jsxProps, key);
};
const cloneElement = (element, props, ...children) => {
  return (0, import_jsx_runtime.jsx)(
    element.tag,
    {
      ...element.props,
      ...props,
      children: children.length ? children : element.children
    },
    element.key
  );
};
var dom_default = {
  useState: import_hooks.useState,
  useEffect: import_hooks.useEffect,
  useRef: import_hooks.useRef,
  useCallback: import_hooks.useCallback,
  use: import_hooks.use,
  startTransition: import_hooks.startTransition,
  useTransition: import_hooks.useTransition,
  useDeferredValue: import_hooks.useDeferredValue,
  startViewTransition: import_hooks.startViewTransition,
  useViewTransition: import_hooks.useViewTransition,
  useMemo: import_hooks.useMemo,
  useLayoutEffect: import_hooks.useLayoutEffect,
  useReducer: import_hooks.useReducer,
  useId: import_hooks.useId,
  useDebugValue: import_hooks.useDebugValue,
  Suspense: import_components.Suspense,
  ErrorBoundary: import_components.ErrorBoundary,
  createContext: import_context2.createContext,
  useContext: import_context.useContext,
  memo: import_base.memo,
  isValidElement: import_base.isValidElement,
  createElement,
  cloneElement
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ErrorBoundary,
  Suspense,
  cloneElement,
  createContext,
  createElement,
  isValidElement,
  jsx,
  memo,
  render,
  startTransition,
  startViewTransition,
  use,
  useCallback,
  useContext,
  useDebugValue,
  useDeferredValue,
  useEffect,
  useId,
  useLayoutEffect,
  useMemo,
  useReducer,
  useRef,
  useState,
  useTransition,
  useViewTransition
});
