// hono/dist/jsx/dom/index.js
// src/jsx/dom/index.ts
import { memo, isValidElement } from "../base.js";
import { useContext } from "../context.js";
import {
  useState,
  useEffect,
  useRef,
  useCallback,
  use,
  startTransition,
  useTransition,
  useDeferredValue,
  startViewTransition,
  useViewTransition,
  useMemo,
  useLayoutEffect,
  useReducer,
  useId,
  useDebugValue
} from "../hooks/index.js";
import { Suspense, ErrorBoundary } from "./components.js";
import { createContext } from "./context.js";
import { jsx } from "./jsx-runtime.js";
import { render } from "./render.js";
var createElement = (tag, props, ...children) => {
  const jsxProps = { ...props, children };
  let key = void 0;
  if ("key" in jsxProps) {
    key = jsxProps.key;
    delete jsxProps.key;
  }
  return jsx(tag, jsxProps, key);
};
var cloneElement = (element, props, ...children) => {
  return jsx(
    element.tag,
    {
      ...element.props,
      ...props,
      children: children.length ? children : element.children
    },
    element.key
  );
};
var dom_default = {
  useState,
  useEffect,
  useRef,
  useCallback,
  use,
  startTransition,
  useTransition,
  useDeferredValue,
  startViewTransition,
  useViewTransition,
  useMemo,
  useLayoutEffect,
  useReducer,
  useId,
  useDebugValue,
  Suspense,
  ErrorBoundary,
  createContext,
  useContext,
  memo,
  isValidElement,
  createElement,
  cloneElement
};
export {
  ErrorBoundary,
  Suspense,
  cloneElement,
  createContext,
  createElement,
  dom_default as default,
  isValidElement,
  createElement as jsx,
  memo,
  render,
  startTransition,
  startViewTransition,
  use,
  useCallback,
  useContext,
  useDebugValue,
  useDeferredValue,
  useEffect,
  useId,
  useLayoutEffect,
  useMemo,
  useReducer,
  useRef,
  useState,
  useTransition,
  useViewTransition
};
